### PR TITLE
Eli/fixed price permit + purchase approval

### LINKED ISSUE #156: Fixed Price Sale with Credit Card Purchases
* `Buy` with permits entrypoint
* Delayed distribution after purchase

### PR DESCRIPTION
**Pull request by EGuenz**
Closes #156 

- Adds fixed price contracts with `Offchain_Buy` entrypoint and `Confirm_purchases`, `Revoke_purchases` entrypoint to be used for off-chain/credit-card purchases. 
- Adds permit ts interface based on work by @michaeljklein as well as integration tests with taquito

### CODE REVIEW

File: packages/minter-contracts/ligo/src/fixed_price_sale/fixed_price_sale_market.mligo [hunk 1]
```diff
@@ -95,24 +110,31 @@ let deposit_for_sale(sale_data, storage: sale_data * storage) : (operation list
...
+    let sale : sale = get_sale(sale_id, storage) in 
+    let token_for_sale_address = sale.sale_data.sale_token.fa2_address in
+    let token_for_sale_token_id = sale.sale_data.sale_token.token_id in
+    let amount_ = sale.sale_data.amount in 
+    let seller = sale.seller in
+    let is_seller = Tezos.sender = seller in
+    let v : unit = if is_seller then ()
+      else fail_if_not_admin_ext (storage.admin, "OR_A_SELLER") in
+#if PERMIT_MARKET
+    let sale : sale = get_sale(sale_id, storage) in
```
Comment (EGuenz) [inline packages/minter-contracts/ligo/src/fixed_price_sale/fixed_price_sale_market.mligo:133]: get sale twice

File: packages/minter-contracts/ligo/src/fixed_price_sale/fixed_price_sale_market_offchain.mligo [hunk 2]
```diff
@@ -0,0 +1,168 @@
+#if !PERMIT_MARKET
+#define PERMIT_MARKET
```
Comment (emishur) [inline packages/minter-contracts/ligo/src/fixed_price_sale/fixed_price_sale_market_offchain.mligo:2]: consider renaming macro to tell **what** functionality is turned on, not **how** it is implemented.
`PERMIT_MARKET` -> `OFFCHAIN_MARKET`

File: packages/minter-contracts/ligo/src/fixed_price_sale/fixed_price_sale_market_offchain.mligo [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+    assert_msg(amount_ >= 1n, "NO_SALE");
+  
+    let new_sales : (sale_id, sale) big_map = 
+      Big_map.update sale_id (Some {sale with pending_purchases = (Set.add purchaser pending_purchases); 
+                                   sale_data = {sale.sale_data with amount = abs (amount_ - 1n)}}) storage.sales in
+    let new_s = {storage with sales = new_sales } in
+    new_s
+  end 
+
+let buy_with_optional_permit (permit_storage, p : permit_storage * permit_buy_param)  : permit_storage =
```
Comment (emishur) [inline packages/minter-contracts/ligo/src/fixed_price_sale/fixed_price_sale_market_offchain.mligo:115]: Shouldn't `_optional_` be removed from the function name?

File: packages/minter-contracts/ligo/src/fixed_price_sale/fixed_price_sale_market_offchain.mligo [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+
+let buy_with_optional_permit (permit_storage, p : permit_storage * permit_buy_param)  : permit_storage = 
+  let param_hash = Crypto.blake2b (Bytes.pack p.sale_id) in 
+  let v : unit = check_permit (p.permit, permit_storage.counter, param_hash) in 
+  let purchaser = address_from_key (p.permit.signerKey) in
+  let market_storage : storage 
+    = buy_token_pending_confirmation (p.sale_id, purchaser, permit_storage.market_storage) in
+  {permit_storage with market_storage = market_storage; counter = permit_storage.counter + 1n}
+
+let buy_with_optional_permits (permits, permit_storage : permit_buy_param list * permit_storage) : permit_return =
```
Comment (emishur) [inline packages/minter-contracts/ligo/src/fixed_price_sale/fixed_price_sale_market_offchain.mligo:123]: Shouldn't `_optional_` be removed from the function name?

File: packages/minter-contracts/ligo/src/fixed_price_sale/fixed_price_sale_market_offchain.mligo (hunk 2 above)
Comment (EGuenz) [reply packages/minter-contracts/ligo/src/fixed_price_sale/fixed_price_sale_market_offchain.mligo]: Fixed thanks!

**Review by emishur — APPROVED** [review]

**Pull request merged**